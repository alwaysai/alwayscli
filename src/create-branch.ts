import { Branch, ExcludeUnderscoreType } from './types';
import { BRANCH } from './constants';

type Config = ExcludeUnderscoreType<Branch>;

export function createBranch(config: Config) {
  const branch: Branch = {
    ...config,
    _type: BRANCH,
  };
  return branch;
}
