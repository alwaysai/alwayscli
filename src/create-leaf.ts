import {
  AnyNamedInputs,
  Leaf,
  ExcludeInternallyAssigned,
  AnyInput,
  Input,
} from './types';
import { LEAF } from './constants';

export function createLeaf<
  T extends AnyInput = Input<undefined, false>,
  U extends AnyNamedInputs = {},
  V extends AnyInput = Input<undefined, false>
>(config: ExcludeInternallyAssigned<Leaf<T, U, V>>) {
  const leaf: Leaf<T, U, V> = {
    ...config,
    _type: LEAF,
  };
  return leaf;
}
