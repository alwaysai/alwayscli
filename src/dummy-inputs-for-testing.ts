import { Input } from './types';

export const DUMMY_INPUT_UNDEFINED_WAS_PASSED = 'undefined was passed';
export const DUMMY_INPUT_EMPTY_ARRAY_WAS_PASSED = 'undefined was passed';
export const DUMMY_INPUT_THROWN_INTENTIONALLY = 'thrown intentionally';
export const DUMMY_INPUT_THROW = 'throw';
export const DUMMY_INPUT_THROW_NON_TRUTHY = 'throw-non-truthy';

export const dummyInput: Input<string, false> = {
  placeholder: '',
  getValue(argv) {
    if (typeof argv === 'undefined') {
      return DUMMY_INPUT_UNDEFINED_WAS_PASSED;
    }
    if (argv.length === 0) {
      return DUMMY_INPUT_EMPTY_ARRAY_WAS_PASSED;
    }
    if (argv[0] === DUMMY_INPUT_THROW) {
      throw new Error(DUMMY_INPUT_THROWN_INTENTIONALLY);
    }
    if (argv[0] === DUMMY_INPUT_THROW_NON_TRUTHY) {
      throw '';
    }
    return argv[0];
  },
  getDescription: () => '',
};

export const dummyRequiredInput: Input<string, true> = {
  placeholder: '<foo>',
  required: true,
  getValue: dummyInput.getValue,
  getDescription: () => '',
};
