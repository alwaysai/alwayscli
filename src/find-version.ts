import readPkgUp = require('read-pkg-up');
import { dirname } from 'path';

export async function findVersion() {
  const mainModule = require.main;
  if (!mainModule) {
    // We not sure under what circumstances require.main would be undefined but
    // the NodeJS types say it might be.
    return undefined;
  }

  const found = await readPkgUp({
    cwd: dirname(mainModule.filename),
    normalize: false,
  });

  if (!found) {
    return undefined;
  }

  if (!found.packageJson.version) {
    return undefined;
  }
  return found.packageJson.version;
}
