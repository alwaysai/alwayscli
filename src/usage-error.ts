export const USAGE = 'USAGE';

export class UsageError extends Error {
  public readonly code: typeof USAGE;
  constructor(message?: string) {
    super(message);
    this.code = USAGE;
    Object.setPrototypeOf(this, new.target.prototype);
  }
}
