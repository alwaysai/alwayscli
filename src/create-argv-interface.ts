import { Leaf, Branch } from './types';
import { accumulateCommandStack } from './accumulate-command-stack';
import { accumulateArgvs } from './accumulate-argvs';
import { accumulateNamedValues } from './accumulate-named-values';
import { UsageError } from './usage-error';
import { TerseError } from './terse-error';
import { BRANCH } from './constants';
import { findVersion } from './find-version';
import { callGetValue } from './call-get-value';
import { LastCommand } from './last-command';

export function createArgvInterface(rootCommand: Branch | Leaf<any, any, any>) {
  return async function argvInterface(argv: string[]) {
    if (['-v', '--version'].includes(argv[0])) {
      const version = await findVersion();
      if (!version) {
        throw new TerseError('Failed to find a "version" string');
      }
      return version;
    }

    const {
      foundHelp,
      commandNamesAndPositionalArgv,
      namedArgvs,
      escapedArgv,
    } = accumulateArgvs(argv);

    const restCommandNamesAndPositionalArgv = accumulateCommandStack(
      rootCommand,
      commandNamesAndPositionalArgv,
    );

    if (foundHelp) {
      // E.g.:
      //   cli branch0 branch1 --help
      // Same as:
      //   cli --help branch0 branch1
      throw new UsageError();
    }

    const lastCommand = LastCommand(rootCommand);

    if (lastCommand._type === BRANCH) {
      if (restCommandNamesAndPositionalArgv[0]) {
        // E.g. cli branch0 branch1 bad-command-name
        throw new UsageError(`Bad command "${restCommandNamesAndPositionalArgv[0]}"`);
      }
      // E.g. cli branch0 branch1
      throw new UsageError();
    }

    let argsValue: any = undefined;
    if (lastCommand.args) {
      // Note that for named and escaped argvs, we distinguish between
      // `undefined` and `[]`. For example, "cli" gives an escaped argv
      // `undefined` whereas "cli --" gives an escaped argv `[]`. For the
      // "positionalArgv", however, there is no such distinction. By convention,
      // we elect here to pass in `undefined` rather than an empty array when we
      // invoke `callGetValue` when `restCommandNamesAndPositionalArgv` is an
      // empty array.
      const positionalArgv =
        restCommandNamesAndPositionalArgv.length > 0
          ? restCommandNamesAndPositionalArgv
          : undefined;
      argsValue = await callGetValue(lastCommand.args, positionalArgv);
    } else if (restCommandNamesAndPositionalArgv.length > 0) {
      throw new UsageError(
        `Unexpected argument "${restCommandNamesAndPositionalArgv[0]}" : Command "${
          lastCommand.name
        }" does not accept positional arguments`,
      );
    }

    const namedValues = await accumulateNamedValues(
      lastCommand.options || {},
      namedArgvs,
    );

    let escapedValue: any = undefined;
    if (lastCommand.escaped) {
      escapedValue = await callGetValue(lastCommand.escaped, escapedArgv, '--');
    } else if (escapedArgv) {
      throw new UsageError(
        `Command "${lastCommand.name}" does not allow "--" as an argument`,
      );
    }

    const result = await lastCommand.action(argsValue, namedValues, escapedValue);
    return result;
  };
}
