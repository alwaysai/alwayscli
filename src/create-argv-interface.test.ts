import { runAndCatch } from '@carnesen/run-and-catch';
import { createBranch } from './create-branch';
import { createLeaf } from './create-leaf';
import { dummyInput } from './dummy-inputs-for-testing';
import { createArgvInterface } from './create-argv-interface';
import { findVersion } from './find-version';
import { USAGE } from './usage-error';

const leafWithNamedInputs = createLeaf({
  name: 'leaf-with-named-inputs',
  options: {
    foo: dummyInput,
  },
  action(...args) {
    return args;
  },
});

const leafWithPositionalInput = createLeaf({
  name: 'leaf-with-positional-input',
  args: dummyInput,
  action(...args) {
    return args;
  },
});

const leafWithEscapedInput = createLeaf({
  name: 'leaf-with-escaped-input',
  escaped: dummyInput,
  action(...args) {
    return args;
  },
});

const root = createBranch({
  name: 'cli',
  subcommands: [leafWithPositionalInput, leafWithNamedInputs, leafWithEscapedInput],
});

const argvInterface = createArgvInterface(root);

describe(createArgvInterface.name, () => {
  it('returns version string from package.json if "-v" or "--version" is passed', async () => {
    const version = await findVersion();
    expect(await argvInterface(['-v'])).toBe(version);
    expect(await argvInterface(['--version'])).toBe(version);
  });

  it('throws USAGE error with empty message if --help is passed', async () => {
    const exception = await runAndCatch(argvInterface, ['--help']);
    expect(exception.code).toBe(USAGE);
    expect(exception.message).toBeFalsy();
  });

  it('throws USAGE error with empty message if last command is a branch and no additional argv is present', async () => {
    const exception = await runAndCatch(argvInterface, []);
    expect(exception.code).toBe(USAGE);
    expect(exception.message).toBeFalsy();
  });

  it('throws USAGE error "bad command" if last command is a branch and additional argv is present', async () => {
    const exception = await runAndCatch(argvInterface, ['oops']);
    expect(exception.code).toBe(USAGE);
    expect(exception.message).toMatch(/bad command/i);
    expect(exception.message).toMatch('"oops"');
    expect(exception.message).toMatchSnapshot();
  });

  it('throws USAGE error "positional arguments" if last command is a leaf without positionalInput property and additional argv is present', async () => {
    const exception = await runAndCatch(argvInterface, [
      leafWithNamedInputs.name,
      'oops',
    ]);
    expect(exception.code).toBe(USAGE);
    expect(exception.message).toMatch('Unexpected argument "oops"');
    expect(exception.message).toMatch(leafWithNamedInputs.name);
    expect(exception.message).toMatch(/positional arguments/i);
    expect(exception.message).toMatchSnapshot();
  });

  it('Passes parsed positional value as first argument of the "action" function', async () => {
    const positionalArgv = ['foo', 'bar'];
    const result = await argvInterface([leafWithPositionalInput.name, ...positionalArgv]);
    expect(result).toEqual([dummyInput.getValue(positionalArgv), {}, undefined]);
  });

  it('Passes parsed named values as second argument of the "action" function', async () => {
    const namedArgv = ['--foo', 'bar'];
    const result = await argvInterface([leafWithNamedInputs.name, ...namedArgv]);
    expect(result).toEqual([undefined, { foo: dummyInput.getValue(['bar']) }, undefined]);
  });

  it(`Throws USAGE error 'does not allow "--"' if leaf does not have an "escaped" property`, async () => {
    const exception = await runAndCatch(argvInterface, [
      leafWithPositionalInput.name,
      '--',
    ]);
    expect(exception.code).toBe(USAGE);
    expect(exception.message).toMatch(leafWithPositionalInput.name);
    expect(exception.message).toMatch('does not allow "--"');
  });

  it('Passes parsed escaped value as third argument of the "action" function', async () => {
    const result = await argvInterface([leafWithEscapedInput.name, '--']);
    expect(result).toEqual([undefined, {}, leafWithEscapedInput.escaped!.getValue([])]);
  });
});
