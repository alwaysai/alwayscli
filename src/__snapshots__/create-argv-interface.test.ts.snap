// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createArgvInterface throws USAGE error "bad command" if last command is a branch and additional argv is present 1`] = `"Bad command \\"oops\\""`;

exports[`createArgvInterface throws USAGE error "positional arguments" if last command is a leaf without positionalInput property and additional argv is present 1`] = `"Unexpected argument \\"oops\\" : Command \\"leaf-with-named-inputs\\" does not accept positional arguments"`;
